demos/guide/tls-client-block.c

```diff
@@ -0,0 +1,252 @@
+/*
+ *  Copyright 2023 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ *  Licensed under the Apache License 2.0 (the "License").  You may not use
+ *  this file except in compliance with the License.  You can obtain a copy
+ *  in the file LICENSE in the source distribution or at
+ *  https://www.openssl.org/source/license.html
+ */
+
+/*
+ * NB: Changes to this file should also be reflected in
+ * doc/man7/guide-tls-client-block.pod
+ */
+
+#include <string.h>
+
+/* Include the appropriate header file for AF_INET and SOCK_STREAM */
+#ifdef _WIN32 /* Windows */
+# include <winsock2.h>
+#else /* Linux/Unix */
+# include <sys/socket.h>
+#endif
+
+#include <openssl/bio.h>
+#include <openssl/ssl.h>
+#include <openssl/err.h>
+
+/* Helper function to create a BIO connected to the server */
+static BIO *create_socket_bio(const char *hostname, const char *port)
+{
+    int sock;
+    BIO_ADDRINFO *ai = NULL;
+    BIO *bio;
+
+    /*
+     * Create a TCP socket. We could equally use non-OpenSSL calls such as
+     * "socket" here for this and the subsequent lookup, connect and close
+     * functions. But for portability reasons and also so that we get errors
+     * on the OpenSSL stack in the event of a failure we use OpenSSL's
+     * versions of these functions.
+     */
+    sock = BIO_socket(AF_INET, SOCK_STREAM, 0, 0);
+    if (sock == -1)
+        return NULL;
+
+    /*
+     * Lookup IP address info for the server. For simplicity we only look at
+     * the first address returned.
+     */
+    if (!BIO_lookup_ex(hostname, port, BIO_LOOKUP_CLIENT, AF_INET,
+                       SOCK_STREAM, 0, &ai)) {
+        BIO_closesocket(sock);
+        return NULL;
+    }
+
+    /* Connect the socket to the server's address */
+    if (!BIO_connect(sock, BIO_ADDRINFO_address(ai), BIO_SOCK_NODELAY)) {
```

Comment:
Perhaps loop through the `ai` list until a connection succeeds, and don't restrict to IPv4?